Beacon diff

leonidc opened the pull request:
This PR is introducing the feature called beacon-diff:

Each Monitor client every 2 seconds sends the beacon with encapsulated GW data: list of subsystems, each subsystem contains list of namespaces and list of listeners.
In case configuration is big this can be pretty big overhead because in the stable site no changes happen - so same big portion of data is accepted by single monitor from many GW monitor clients.
In order to improve this behavior we designed the some kind of protocol:

In case Monitor client upon sending the beacon sees no difference on the subsystem, then this subsystem and its encapsulated "child's"(namespaces, listeners) are not present in the beacon
In case some subsystem changed from the previous beacon, it is added to the beacon alone with its child's, also added change-descriptor that describes the change (3 possible descriptor values ADDED, REMOVED, CHANGED)
Sequence number is added to each beacon. In case monitor gets beacon out of order, it ignores it and signals to the Monitor client to "return back" to the missing sequence number that monitor sends to it in the beacon ACK
When Monitor client receives ACK with indication of sequence out-of-order, it sends all subsystems in the next beacon(beacon diff is not calculated for the next beacon), it also sends the sequence number expected by the monitor.
see also
https://tracker.ceph.com/issues/72394


## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

leonidc committed:
nvmeofgw: beacon diff implementation in the monitor and in the MonClient.

     -monclient encodes subsystems by beacon-diff rules if BEACON_DIFF
      bit is enabled by quorum
     -monitor processes beacons by beacon-diff new schema
     -monitor detects sequence out of order(ooo) condition and handles it
     -in case ooo detected monitor send ack to the gw with the expected correct sequence
     -monitor skips failovers for some interval when ooo detected
     -monitor ignores all becons with incorrect sequences until gw sends expected one
     -coding upgrade rules

  Signed-off-by: Leonid Chernin <leonidc@il.ibm.com>

Fixes: https://tracker.ceph.com/issues/72394

ronen-fr commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -316,7 +317,9 @@ inline  void decode(
   uint8_t version = 1;
-
+  if (HAVE_FEATURE(features, NVMEOF_BEACON_DIFF)) {
+     version = 2;
```
A number ('2') here, but a constant on line 858?

leonidc commented on src/nvmeof/NVMeofGwMonitorClient.cc:
```diff
@@ -252,7 +252,10 @@ void NVMeofGwMonitorClient::send_beacon()
-  bool include_diff = true;
+  bool include_diff = HAVE_FEATURE(cluster_features, NVMEOF_BEACON_DIFF);
+
+  dout(10) << "Cluster features: 0x" << std::hex << cluster_features << std::dec
```
done

ronen-fr replied:
BTW - '{:#x}' (adding the '#' sign) adds the '0x' automagically.

caroav commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -19,6 +19,11 @@
 
+// Local version constants that match MNVMeofGwBeacon.h
+// We can't include that header here due to circular dependencies
+#define BEACON_SUBSYS_VERSION_LEGACY 1      // Legacy beacon format (no diff support)
```
@baum can you address this comment?

leonidc replied:
@baum , done

leonidc commented:
This  PR is introducing the feature called  beacon-diff:

Each Monitor client every 2 seconds sends  the beacon with encapsulated GW data:  list of subsystems, namespaces, listeners 
In case configuration is big  this can be pretty big overhead  because  in the stable site no changes  happen - so  same big portion of data is accepted  by single monitor  from many GW monitor clients. 
In order to improve this behavior we designed  the  some kind of protocol: 
- In case  Monitor client  upon sending the beacon  sees no difference on the subsystem, then this subsystem and its encapsulated "child's"(namespaces, listeners) are not present in the beacon
-In case some subsystem  changed from the previous beacon, than it is added to the beacon alone with its child's, also added change-descriptor that describes the change (3 possible descriptor values  ADDED, REMOVED, CHANGED)
-Sequence number is added to each beacon. In case monitor gets beacon out of order, it ignores it and signals to the Monitor client to "return back"  to  the missing number  that monitor sends to it  in the beacon ACK

ronen-fr commented:
> @ronen-fr can you check if all of your comments were addressed? Thx.

Not the first one:
"Please add a PR comment, describing the highlights."

leonidc commented:
I just squashed all 6 commits to 1, added a tracker and re-based the PR

athanatos commented on src/include/ceph_features.h:
```diff
@@ -162,6 +163,7 @@ DEFINE_CEPH_FEATURE(49, 2, SERVER_SQUID);
 DEFINE_CEPH_FEATURE_RETIRED(50, 1, MON_METADATA, MIMIC, OCTOPUS)
 DEFINE_CEPH_FEATURE(50, 2, SERVER_TENTACLE);
 DEFINE_CEPH_FEATURE_RETIRED(51, 1, OSD_BITWISE_HOBJ_SORT, MIMIC, OCTOPUS)
+DEFINE_CEPH_FEATURE(51, 2, NVMEOF_BEACON_DIFF)
```
I see the same feature defined twice with different names -- why?

leonidc replied:
@athanatos , thanks will delete  retired feature line

leonidc replied:
@gregsfortytwo , please see now

athanatos reviewed (changes requested):
See above comment, I still need to review the rest of the PR.

gregsfortytwo reviewed (changes requested):
I didn't review the whole thing, but I have some concerns in the common core areas I looked at.

athanatos commented on src/mon/NVMeofGwMap.h:
```diff
@@ -65,15 +65,17 @@ class NVMeofGwMap
     epoch_t& last_osd_epoch,  bool &propose_pending);
   int process_gw_map_gw_down(
-    const NvmeGwId &gw_id, const NvmeGroupKey& group_key,
+    const NvmeGwId &gw_id, const Nv/meGroupKey& group_key,
```
Does this compile?  Nv/meGroupKey looks like a typo.

leonidc replied:
thanks @athanatos , typo

athanatos commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -824,22 +850,41 @@ inline void decode(BeaconListener& ls, ceph::buffer::list::const_iterator &bl) {
-inline void encode(const BeaconSubsystem& sub,  ceph::bufferlist &bl) {
-  ENCODE_START(1, 1, bl);
+inline void encode(const BeaconSubsystem& sub,  ceph::bufferlist &bl, uint64_t features) {
+  uint8_t version = BEACON_VERSION_LEGACY;  // Default to legacy version
```
If you want to use a common pair of version symbols (BEACON_VERSION_LEGACY and BEACON_VERSION_ENHANCED) for these data structures, that's reasonable, but please carefully audit all of the relevant data structures and make sure you aren't using 1 or 2 to check versions.  Please also list the data structures those versions apply to where the two symbols are defined.  I disagree with @ronen-fr 's request that you move them out of this file -- the message header needs to include this file anyway, and those two symbols are closely tied to the data structures in this file.  This file is therefore a good place for them.

Alternately, you can drop them and simply let the individual encode/decode pairs use bare, unrelated version numbers as in the rest of the code base.

athanatos commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -50,11 +51,16 @@ class MNVMeofGwBeacon final : public PaxosServiceMessage {
+        uint64_t sequence_ = 0,  // default sequence for backward compatibility
+        bool enable_diff = false)  // default to legacy behavior for backward compatibility
+    : PaxosServiceMessage{MSG_MNVMEOF_GW_BEACON,
+                          static_cast<version_t>(enable_diff ? 1 : 0), // user_version: 1=enhanced, 0=legacy
```
This version field is a paxos version field and is unrelated to the encoding version of the message.  I *think* this should be set to 0 since your service does not care about the paxos readable version check -- @gregsfortytwo does that sound right?

leonidc replied:
@athanatos , remember  what we agreed  with you that  Monitor  distinguishes  coding of beacon by the version sent by Monclient.  Because Monclient does not have access to the quorum features - only connection features.
So in the current concept if upon upgrade "new" Monclient  sends beacon to the "new" monitor but quorum still not passed to new features , monlient enables beacondiff feature and monitor successfully decodes it   and locally enables beacon diff. but monitor's map sent to the peons is guided by the quorum features

leonidc replied:
@athanatos  , ah ok , you are right
there are 2 parameters called version

leonidc commented on src/include/ceph_features.h:
```diff
@@ -149,6 +149,7 @@ DEFINE_CEPH_FEATURE(43, 1, OSD_POOLRESEND)   // 4.13
 DEFINE_CEPH_FEATURE(44, 2, NVMEOFHA)
 DEFINE_CEPH_FEATURE_RETIRED(45, 1, OSD_SET_ALLOC_HINT, JEWEL, LUMINOUS)
 DEFINE_CEPH_FEATURE(45, 2, NVMEOFHAMAP)
+DEFINE_CEPH_FEATURE(51, 2, NVMEOF_BEACONDIFF)
```
ok thanks

leonidc commented on src/mon/NVMeofGwMap.cc:
```diff
@@ -82,10 +83,10 @@ void NVMeofGwMap::remove_grp_id(
 {
   std::set<NvmeAnaGrpId> allocated;
-  if (HAVE_FEATURE(mon->get_quorum_con_features(), NVMEOFHAMAP)) {
+  if (test || HAVE_FEATURE(mon->get_quorum_con_features(), NVMEOFHAMAP)) {
```
The motivation for this change:
for the beacon diff feature  we  regain  the testing infrastructure. But test using cfg_create_gw/cfg_delete_gw  crashed  because of access to mon . In order to prevent this the new parameter bool "test" was added

leonidc commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -92,10 +99,14 @@ class MNVMeofGwBeacon final : public PaxosServiceMessage {
     encode(gw_pool, payload);
     encode(gw_group, payload);
-    encode(subsystems, payload);
+    encode(subsystems, payload, features);
```
This message is sent from the GWs MonClient to the NVMeofMonitor . If Client reading the  connection features  enabled  the beacon diff it encodes subsystems by the new rules , it also increments the version. 
Monitor  decodes correctly the subsystems guided by the version of the  beacon, even when quorum not passed to the last version supporting the beacon-diff. This was an idea, does it make sense @gregsfortytwo ?

leonidc replied:
ok, @athanatos , @gregsfortytwo , thanks, resolving this comment

Jayaprakash-ibm commented on src/test/test_nvmeof_mon_encoding.cc:
```diff
@@ -128,10 +138,10 @@ void test_MNVMeofGwMap() {
-  pending_map.cfg_delete_gw("GW3" ,group_key);
-  dout(0) << "deleted GW3 . we should see the empty map " << pending_map << dendl;
+  //pending_map.cfg_delete_gw("wow" ,group_key, true);
+/
```
Hi @leonidc , This line here is causing jenkins failures, remove this `/`
please check : https://jenkins.ceph.com/job/ceph-dev-pipeline/1540/
https://jenkins.ceph.com/job/ceph-dev-pipeline/1540/consoleFull#-5911579562a811ea2-3e7b-466b-84b4-d13df7e35809

leonidc commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -36,10 +35,12 @@ class MNVMeofGwBeacon final : public PaxosServiceMessage {
   MNVMeofGwBeacon()
-    : PaxosServiceMessage{MSG_MNVMEOF_GW_BEACON, 0, HEAD_VERSION, COMPAT_VERSION}
+    : PaxosServiceMessage{MSG_MNVMEOF_GW_BEACON, 0, BEACON_VERSION_ENHANCED,
+      BEACON_VERSION_LEGACY}, sequence(0)
```
@athanatos, @gregsfortytwo   this  a code of the default constructor 
Actually  MonitorClient  before sending the beacon  gets features from its connection with the monitor.
If it  is connected to the new monitor it encodes BEACON_VERSION_ENHANCED  in the version of the beacon.

Compatible version always set to 1 - unconditionally. We tried different  combinations of old/new client-monitor during upgrade and it seems works well.
In case  "ENHANCED" MonitorClient  connected to the "old" monitor - MonitorClient formats the LEGACY message - old beacon format is sent to it - nothing changed for the monitor
So old monitor never receives ENHANCED beacon !

gregsfortytwo replied:
It is possible you manage to avoid actually sending a badly-encoded message to old code, but you are breaking the rules of Ceph struct_v encoding by pretending that a newly-encoded message can be understood by old code — that is the statement that you are making by not incrementing compat_v. You need to follow the rules so other people can understand it, and because there will be other situations where it doesn't work out so well for you.

And as we've said here and elsewhere, it's really not appropriate to tie message encoding entirely to the map encoding you're using throughout the system.

leonidc commented on src/msg/Message.cc:
```diff
@@ -897,7 +897,7 @@ Message *decode_message(CephContext *cct,
   case MSG_MNVMEOF_GW_BEACON:
     m = make_message<MNVMeofGwBeacon>();
-  break;
+    break;
```
thanks @baum  , marking as resolved

leonidc commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -308,23 +314,37 @@ inline  void decode(
-  ENCODE_START(1, 1, bl);
+  uint8_t version = 1;
+  if (HAVE_FEATURE(features, NVMEOF_BEACON_DIFF)) {
+     version = BEACON_VERSION_ENHANCED;
```
Done, thanks @athanatos

athanatos commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -92,10 +101,14 @@ class MNVMeofGwBeacon final : public PaxosServiceMessage {
     encode(last_osd_epoch, payload);
     encode(last_gwmap_epoch, payload);
+    // Only encode sequence for enhanced beacons (version >= HEAD_VERSION)
+    if (get_header().version >= HEAD_VERSION) {
```
First, this should be >= ENHANCED.  If you bump the version again, HEAD_VERSION will no longer be the correct check.

Second, COMPAT_VERSION is always set to LEGACY.  By doing that, you are stating that clients at LEGACY can skip this field no problem.  The testing you mention in https://github.com/ceph/ceph/pull/65563#discussion_r2454003001 suggests that you are both a) already sending ENHANCED encodings to old clients and that b) the clients deal with them fine (which is expected -- the encoding header includes a length so that old decoders can simply skip the fields they don't recognize).  I only expect to see conditional encodings like this when the client can't actually deal with them.  Why is this necessary?

athanatos replied:
Do change HEAD_VERSION to ENCHANCED if you choose to keep this.

athanatos commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -308,23 +313,35 @@ inline  void decode(
+  if (version >= 2) {
+    encode((uint64_t)state.last_beacon_seq_number, bl);
+    encode(state.last_beacon_seq_ooo, bl);
+  }
```
This encoding a) sets the encode version depending on features and b) sets COMPAT to match version.  I'm not sure you actually need to bother.  You are simply appending two fields, so unconditionally encoding at ENHANCED with LEGACY as COMPAT should be fine.  Is there a specific reason you need to avoid sending ENHANCED to a LEGACY client in this case?

athanatos commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -308,23 +313,35 @@ inline  void decode(
   uint32_t avail;
   decode(avail, bl);
   state.availability = (gw_availability_t)avail;
+  if (struct_v >= 2) {
```
This check is both correct and necessary, regardless of whether encode includes these fields conditionally.

athanatos commented on src/mon/NVMeofGwSerialize.h:
```diff
@@ -824,22 +847,41 @@ inline void decode(BeaconListener& ls, ceph::buffer::list::const_iterator &bl) {
   for (const auto& ns: sub.namespaces)
     encode(ns, bl);
+  if (version >= 2) {
+    encode((uint32_t)sub.change_descriptor, bl);
```
This is another example where you are simply appending a field, I don't think you need a conditional encoding here, you could simply unconditionally encode at version 2.

gregsfortytwo commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -50,11 +51,16 @@ class MNVMeofGwBeacon final : public PaxosServiceMessage {
+    : PaxosServiceMessage{MSG_MNVMEOF_GW_BEACON,
+                          static_cast<version_t>(enable_diff ? 1 : 0), // user_version: 1=enhanced, 0=legacy
+                          enable_diff ? BEACON_VERSION_ENHANCED :
+                          BEACON_VERSION_LEGACY, BEACON_VERSION_LEGACY},// Minimum compatible version
```
Something in your logic is broken if updating the compat version caused the monitors to crash. You'll need to understand what happened before we can merge this PR. What was the monitor crash?
I do not believe you can make the changes you are making here without updating the compat version, but you may be able to convince me if you can walk through why it's safe.

athanatos commented on src/nvmeof/NVMeofGwMonitorClient.cc:
```diff
@@ -231,26 +236,45 @@ void NVMeofGwMonitorClient::send_beacon()
       pool,
       group,
-      subs,
+      include_diff ? subs : current_subsystems,
```
cluster_features isn't a good idea.  You don't need to know (and here, you actually do not know) the features for the monitor quorum, you only need the features of the specific monitor you are sending the message to.  encode_payload is always passed the feature vector for the target of the message.  Use that.  I don't think the constructor here needs to know at all.  encode_payload can set header.version to an older version if necessary, though I don't think you even need to do that (see my comment on encode_payload).

leonidc replied:
@athanatos , cluster_feature is  confusing name of the variable. It actually gets connection features. See the code:
  
  uint64_t features = m->get_connection()->get_features();
  cluster_features |= features;

athanatos replied:
@gregsfortytwo and I are investigating the right way to handle this.  We'll let you know.

gregsfortytwo replied:
Yes, that commit looks like the right set of places to poke.
I did mention (just want to preserve the comment in case of rebase) that I think we should just map the CEPH_MON_FEATURE_INCOMPAT_NVMEOF_BEACON_DIFF value to TENTACLE+1 rather than to 32 (skipping a bunch).

gregsfortytwo reviewed (dismissed):
Sam is doing better review, but here's a few things...
The overall theme here is that if you're going to play with this code, you need to understand Ceph's versioning mechanisms and work with them properly. We've discussed them a lot in PRs; would it help if we set up a presentation to go over it with everybody?

athanatos committed:
mon: add NVMEOF_BEACON_DIFF to mon_feature_t and mon CompatSet

NOPE NOPE
In order for the client to safely send BEACON_DIFF messages, it
needs to be the case that the leader at the time of receipt will
support BEACON_DIFF.

Simply using the connection features for the MonClient's target mon is
insufficient, because it might be a peon.  If the peon supports
BEACON_DIFF and the leader does not the leader will either crash or
interpret it as a full BEACON.  Neither outcome is acceptable.

Instead, we need to wire up a feature bit to the MonMap mon_feature_t
members and the CompatSet.

Adding FEATURE_BEACON_DIFF to ceph::features::mon get_supported()
and get_persistent() ensures that once all monitors in the quorum
support it, MonMap::get_required_features() will include it.
See Elector::propose_to_peers, Monitor::(win|lose)_election,
MonmapMonitor::apply_mon_features.

Once FEATURE_BEACON_DIFF is present in MonMap::get_required_features():
- Monitor::apply_monmap_to_compatset_features() will prevent
  downgrades of the monitors by updating the CompatSet to include
  CEPH_MON_FEATURE_INCOMPAT_NVMEOF_BEACON_DIFF
- Monitor::calc_quorum_requirements() will set
  Monitor::required_features to require the NVMEOF_BEACON_DIFF
  for any monitor peers.
- MonClient::get_monmap_required_features() will eventually include
  ceph::features::mon::FEATURE_NVMEOF_BEACON_DIFF.

Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos committed:
mon/MonClient: add get_monmap_required_features() helper

Signed-off-by: Samuel Just <sjust@redhat.com>

athanatos commented on src/messages/MNVMeofGwBeacon.h:
```diff
@@ -22,11 +22,12 @@
-  static constexpr int HEAD_VERSION = 1;
-  static constexpr int COMPAT_VERSION = 1;
+  static constexpr int ENHANCED_VERSION = BEACON_VERSION_ENHANCED;
+  static constexpr int COMPAT_VERSION = BEACON_VERSION_LEGACY;
```
Why introduce new names for BEACON_VERSION_ENHANCED and BEACON_VERSION_LEGACY?  Just use them directly.  The HEAD/COMPAT pattern used by other messages simply isn't useful here because it's not safe for a LEGACY client to decode an ENHANCED message.

leonidc replied:
done , thanks

athanatos reviewed (changes requested):
Let me know if you need more information about how to encode subsystems in MNVMeofGwBeacon.  It's going to be complex going forward as well, so it's important to understand it thoroughly.

Also, please update the commits as you want them in the final merge before requesting another review.

athanatos reviewed (changes requested):
Fix commits.

athanatos reviewed (commented):
LGTM, I'll approve after a qa run.

leonidc committed:
nvmeofgw: prevent map corruption while processing beacons from deleted gws

Fix race issue of map corruption when deleted gw sends beacons
but this gw data was removed from pending map and still exists in map.
Process beacons only if GW's data exists in both maps:
main-map and pending-map, otherwise just ignore beacons.

fixes: https://tracker.ceph.com/issues/74160

Signed-off-by: Leonid Chernin <leonidc@il.ibm.com>

kshtsk commented:
Nice, another issue https://tracker.ceph.com/issues/74846
